Use Cockpit's test images and infrastructure

larskarlitski opened the pull request:
Allows to run the tests on multiple operating systems and on the
infrastructure that the Cockpit team maintains.

`make vm` downloads one of Cockpit's test images (override which one
with TEST_OS) and installs a local build of lorax rpms. The resulting
image is placed in `test/images/$TEST_OS`.

TEST_OS can be set to any of Cockpit's supported images (default:
fedora-30).

Run `make check-vm` to run the CLI checks in the VM. The bulk of the
work is done in `test/check-cli`, which uses Cockpit's `bots` library to
start the VM and run the script in it.

Also included is a `test/run` script, which is the entrypoint for
Cockpit's test infrastructure.


--- Merge policy ---

- [ ] Travis CI PASS
- [ ] `*-aws-runtest` PASS
- [ ] `*-azure-runtest` PASS
- [ ] `*-images-runtest` PASS
- [ ] `*-openstack-runtest` PASS
- [ ] `*-vmware-runtest` PASS
- [ ] For `rhel8-*` and `rhel7-*` branches commit log references an approved
  bug in Bugzilla. Do not merge if the bug doesn't have the 3 ACKs set to `+`!

martinpitt commented on test/check-cli:
```diff
@@ -0,0 +1,44 @@
+
+    r = subprocess.run(["ssh", "-o", "ControlPath=" + machine.ssh_master,
+                               "-p", machine.ssh_port,
+                               machine.ssh_user + "@" + machine.ssh_address,
```
I suggest to assemble the ssh argv prefix above,  and re-use it both here and in the print (but I suppose the print is also just an intermediate debugging thing)

larskarlitski replied:
Indeed, I had used it for debugging. It might be helpful in the future though. I've moved the common prefix out and printed a message with it. Thanks!

atodorov commented on Makefile:
```diff
@@ -12,6 +12,12 @@ TAG = lorax-$(VERSION)-$(RELEASE)
 IMAGE_RELEASE = $(shell awk -F: '/FROM/ { print $$2}' Dockerfile.test)
 
+ifeq ($(TEST_OS),)
+TEST_OS = fedora-30
```
Does the CI infrastructure defined TEST_OS for the various RHEL builds? 

Is it possible we have access to the latest RHEL snapshots/content repositories ?

larskarlitski replied:
Yes. We're installing and subscribing those VMs in [rhel.setup]( https://github.com/cockpit-project/cockpit/blob/master/bots/images/scripts/rhel.setup). Cockpit only subscribes them during installation. For Composer, we probably want to resubscribe them to test pulling content from those repos. 

You can see all available images in Cockpit's [image-refresh](https://github.com/cockpit-project/cockpit/blob/master/bots/image-refresh#L28) script.

atodorov commented on test/vm.install:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh -eu
+
+rpm -e lorax lorax-templates-generic lorax-lmc-virt lorax-lmc-novirt lorax-composer composer-cli || true
```
For composer we'll need the ability to install an arbitrary list of RPMs (URLs to RPMs more precisely). This will facilitate testing pre-built packages from Brew/RHEL snapshots and also allow upgrading/installing other packages as necessary - for example anaconda RPMs to test changes in anaconda before they are merged. 

At the moment Jenkins supports the `EXTRA_RPMS` variable which contains all of these URLs and the system provisioning takes care to download them and install them onto the system before testing.

larskarlitski replied:
`vm.install` installs the rpms that were just built in the source directory. We could add something like `EXTRA_RPMS` here, but I'm afraid that conflating these two use cases in one script might make things complicated (for example, we must ensure that the right versions of the packages are installed). It might be simpler to have this in two scripts.

atodorov commented on test/check-cli:
```diff
@@ -0,0 +1,45 @@
+    print(" ".join(ssh_command))
+    print()
+
+    r = subprocess.run(ssh_command + ["CLI=/usr/bin/composer-cli", "/tests/test_cli.sh"])
```
or maybe `make test_images test_aws test_azure`, etc.

All of these require credentials for the various cloud providers. Is it possible to securely store these credentials in the new CI infrastructure? 

More importantly quickly updating the credentials (also via ENV/manually) when they change or when we want to execute tests with a different account ?

larskarlitski replied:
Let's call the test scripts directly from here, as this script is already called from the Makefile. Too many circles :)

Yes, we can store credentials in the CI infrastructure. We already do this for GitHub tokens and the RHEL subscription. Are you ok to get the basic machinery working in this PR with only the tests that run without extra credentials. Then do a follow up with one that adds support for the cloud tests?

larskarlitski commented:
What do you think, should we merge `test` and `tests`? The entry point for Cockpit's infrastructure is `test/run`, so we're at least stuck with that.

bcl commented on test/vm.install:
```diff
@@ -0,0 +1,10 @@
+
+sudo dnf install -y beakerlib /var/tmp/*.rpm
+
+systemctl enable lorax-composer.socket
```
Since this is a test we may as well just start up lorax-composer.service instead, may save a small amount of time waiting for the metadata to be fetched.

larskarlitski replied:
This script is only run once to install the virtual machine. Starting the service at this stage won't make a difference. (The VM is then booted for every test.)

bcl replied:
If we use lorax-composer.service instead it will start at boot time when it reboots.
Enabling the  .socket is really most useful for systems where you may or many not be using it right away.

larskarlitski replied:
Ah sorry, I misunderstood (you probably didn't mean that, but wrote "_start_ up..." instead of "enable ..."- systemd nomenclature is weird).

As you say, it will probably only make a small difference, but it's cleaner indeed. Thanks!

martinpitt replied:
Please do whatever happens on a production system. It doesn't hurt to test that the socket activation (including the delay on the first call) works as expected.

larskarlitski replied:
Good point. Left it as is.

atodorov commented on test/check-cli:
```diff
@@ -0,0 +1,45 @@
+    print(" ".join(ssh_command))
+    print()
+
+    r = subprocess.run(ssh_command + ["CLI=/usr/bin/composer-cli", "/tests/test_cli.sh"])
```
Add 3rd parameter the path to the test you want to execute. Probably remove the target from Makefile ?

Will require `./$TEST` to change to `$TEST` inside test_cli.sh to account for the fact that you are passing absolute path. I've tried it locally and it worked for me.

larskarlitski replied:
I converted `test/check-cli` to use `unittest` and included all scripts under `tests/cli`. I've also changed `./$TEST` to `$TEST`. I left the `make` targets in there, because you said you sometimes manually run the tests for different rpms?

We probably want to run some of the tests in `/tests/composer` in virtual machines as well.

atodorov commented on test/check-cli:
```diff
@@ -0,0 +1,50 @@
+        self.runCliTest("/tests/cli/test_compose_sanity.sh")
+
+
+class TestImages(TestCli):
```
This will also inherit the `test*Sanity` methods and execute them when this class is executed, which will add more time to the execution duration. Probably not what you want.

IMO move `runCliTest()` into `ComposerTestCase` class.

larskarlitski replied:
Oops indeed. Changed it back. Thanks!

atodorov commented on test/check-cli:
```diff
@@ -0,0 +1,50 @@
+        self.runCliTest("/tests/cli/test_compose_live-iso.sh")
+
+
+class TestCloud(TestCli):
```
Just a note at this point:
here all of the test methods will be executed in sequence and uploads (especially for Azure) tend to take a lot of time. We want to execute all of the cloud methods in parallel, and have them report separate statuses via the GitHub API. This we can deal with later.

Execution wise all of these need a VM just to build the image and then everything else happens on the cloud vendor side.

larskarlitski replied:
Yeah... I've moved them into a separate file and made the test being run dependent on `TEST_SCENARIO`, which is set by Cockpit's test framework to get separate GitHub statusses.

atodorov commented on test/composertest.py:
```diff
@@ -0,0 +1,55 @@
+        return subprocess.run(self.ssh_command + command, **args)
+
+
+def main():
```
That should be `if __name__ == "__main__":`

larskarlitski replied:
No, this script is not meant to be run by itself. The `main` method is used by the `check-*` scripts. It will grow command line options in the future, like `--sit`.

atodorov commented:
@larskarlitski - IMO you can move everything from `test` to `tests` but that will make the PR very big

larskarlitski commented:
Update:
* Resize the image to make room for a composer-generated one
* Add `test/README.md`
* incorporate Alex' suggestions (thanks!)

> @larskarlitski - IMO you can move everything from `tests` to `test` but that will make the PR very big

Let's do this later to keep this easy to review.

martinpitt commented:
@larskarlitski : Seems we need a few more packages in our cockpit/tasks container for building?

larskarlitski commented:
> @larskarlitski : Seems we need a few more packages in our cockpit/tasks container for building?

Yeah. How do I go about doing that?

martinpitt commented:
@larskarlitski : They need to be added to https://github.com/cockpit-project/cockpituous/blob/master/tasks/Dockerfile , the container rebuilt and pushed (`sudo make tasks-container tasks-push`). Then we restart the bots.

martinpitt commented:
@larskarlitski : New containers are rolled out, so this is the time to try again.

martinpitt commented:
The travis errors also seem relevant.

larskarlitski commented:
> The travis errors also seem relevant.

Indeed! One error wasn't. I've put that in a separate commit.

larskarlitski commented:
Another update: travis was still failing because `test` does exist now and the `test` make target was not in `.PHONY`.

Also remove the hack around resizing the image manually now. https://github.com/cockpit-project/cockpit/pull/11681 is merged, so we can use `image-customize --resize` now. (The hack didn't work on Cockpit's infrastructure, because it didn't take `TEST_DATA` into account.)

larskarlitski commented:
```
- nothing provides python(abi) = 3.6 needed by composer-cli-31.3-1.fc28.x86_64
```
@pitti Looks like we want to build the rpms inside of the VM? (Fedora 30 has python 3.7, but the build container only 3.6.)

The remaining travis errors seem unrelated. @atodorov do they work on master?

atodorov commented on test/README.md:
```diff
@@ -0,0 +1,56 @@
+    $ make vm
+
+This downloads a base image from Cockpit's infrastructure. You can control
+which image is downloaded with the `TEST_OS` environment variable. It then
```
Link or examples of acceptable values would be nice

martinpitt replied:
https://github.com/cockpit-project/cockpit/blob/master/test/README.md#test-configuration

larskarlitski replied:
Good point, thanks. I've added this link.

atodorov commented on test/README.md:
```diff
@@ -0,0 +1,56 @@
+
+This downloads a base image from Cockpit's infrastructure. You can control
+which image is downloaded with the `TEST_OS` environment variable. It then
+creates a new image based one that (a qemu snapshot) in `tests/images`, which
```
typo: one -> on

larskarlitski replied:
:+1:

atodorov commented on test/run:
```diff
@@ -0,0 +1,11 @@
+make vm
+
+if [ -n "$TEST_SCENARIO" ]; then
+  test/check-cloud TestCloud.test_$TEST_SCENARIO
```
Can $TEST_SCENARIO actually be defined as Class.test_Method? Then we can move all of the runCliTest methods in one file and be able to start whichever scenario we want. Sort of treating the new Python unittest files like a substitute for Makefile.

larskarlitski replied:
I think this is going a bit far. The test runners won't run all those jobs in parallel anyway. Are you ok with going forward like this and changing it if we notice problems?

atodorov replied:
+1 for merging as-is and I can tweak the rest later.

atodorov commented:
> The remaining travis errors seem unrelated. @atodorov do they work on master?

No, master is broken in Travis CI. I think this has to do with recent rpm changes which broke rpmfluff, which Brian mentioned some time ago.

larskarlitski commented:
Sorry, I didn't have time to work on this recently (vacation and Red Hat Summit).

Update: As I mentioned in a previous comment, we need to build the packages on the test VM itself to get the packages to depend on the right python version. I've changed `vm.install` to build them in a mock. Alas, this makes `make vm` a bit slower, because mock takes some time to unpack/pack the build root.

On first run, I see this error from mock: @martinpitt, do you know what could be going on here?
```
Warning: Module or Group 'buildsys-build' does not exist.
```

martinpitt commented:
> Alas, this makes make vm a bit slower, because mock takes some time to unpack/pack the build root.

Our test images have a pre-unpacked mock chroot in /var/lib/mock. You should be able to run

    su builder -c "/usr/bin/mock --offline --no-clean --shell"

without much delay (< 1 s). Same with building the actual package, of course. For building lorax you probably need to drop  `--offline` so that it can install new build deps; we use that for Cockpit as we know/want to ensure everything is pre-installed. Once that works, we can of course do that for lorax too, to speed up the tests.

martinpitt commented on Makefile:
```diff
@@ -95,12 +101,27 @@ set-docs-owner:
 dist: tag archive
 	scp $(PKGNAME)-$(VERSION).tar.gz fedorahosted.org:lorax
 
+rpm: archive $(PKGNAME).spec
```
Is that still an actually useful target? Or just for developer convenience? If the former, please add it to `.PHONY`.

martinpitt commented on test/vm.install:
```diff
@@ -0,0 +1,34 @@
+
+tmpdir=$(mktemp -d $PWD/srpm-build.XXXXXX)
+tar -xaf $TEST_SOURCE -O $(basename --suffix .tar.gz $TEST_SOURCE)/lorax.spec > $tmpdir/lorax.spec
+rpmbuild -bs \
```
FTR, the *source* rpm can easily be built in the container, outside of the VM, like we do in starter-kit and friends. This works as well, but the .srpm is an useful developer artifact to have outside, for copr, mock builds, inspection, etc.

martinpitt commented on test/vm.install:
```diff
@@ -0,0 +1,34 @@
+rm -rf $tmpdir
+
+mock --installdeps $srpm
+mock --offline --resultdir build-results --rebuild $srpm
```
Please run these two as user "builder". This may explain the permission errors you saw.

larskarlitski commented:
Thanks Martin! I've changed it to build the srpm in the container / on the host (and added `srpm` to `.PHONY`) and am running mock as `builder` now. This makes `vm.install` much cleaner.

martinpitt commented on Makefile:
```diff
@@ -129,8 +141,34 @@ docs-in-docker:
 # nothing to do here, but Jenkins expects this to be present, otherwise fails
 
-.PHONY: docs
+.PHONY: docs check test srpm
```
plus `vm vm-reset`

larskarlitski replied:
Thanks!

martinpitt commented on Makefile:
```diff
@@ -129,8 +141,34 @@ docs-in-docker:
 ci: check test
 
+$(VM_IMAGE): TAG=HEAD
+$(VM_IMAGE): archive bots
```
I think this needs to depend on `srpm`, not `archive`?

larskarlitski replied:
Yeah, indeed. Worked for me because I didn't delete it between runs.

martinpitt commented:
[last run](https://209.132.184.41:8493/logs/pull-689-20190517-100707-cc6a6561-cockpit-fedora-30/log) failed with an ENOSPC, investigating whether it's a problem inside the VM (needs to grow larger?) or outside in the infra.

martinpitt commented:
This PR is alright for the test integration changes now. But it still fails due to ENOSPC, the test should be fixed to clean up built images after it's done verifying them. Otherwise the demands on the test beds are just unreasonably high (also for developers running the test locally).

martinpitt commented:
@larskarlitski : FYI, these are the two commands to run in the 
[webhook container](https://console.apps.ci.centos.org:8443/console/project/cockpit/browse/pods/webhook-mwqw2?tab=terminal) to trigger a test:
```
bots/tests-trigger --repo weldr/lorax 689 cockpit/fedora-30
bots/tests-scan --repo weldr/lorax -p 689 -c cockpit/fedora-30 --amqp amqp.cockpit.svc.cluster.local:5671
```

larskarlitski commented:
Added a few commits to fix tar and qcow2 tests and to make the log more readable (no `tar -v`, set `TEST` so that the test name appears in the beakerlib summary, and incrrease `unittest`'s verbosity.

martinpitt commented on test/vm.install:
```diff
@@ -0,0 +1,18 @@
+# Grow root partition to make room for images. This only works on Fedora right now.
+echo ", +" | sfdisk -N 2 -f /dev/vda
+partprobe
+pvresize /dev/vda2
```
This works on RHEL images, too, as "vda2" is defined by the outside image build process and QEMU/libvirt invocation. If you want to keep it a bit more flexible, use 

    pvs --noheadings -opv_name | xargs pvresize

martinpitt commented on test/vm.install:
```diff
@@ -0,0 +1,18 @@
+echo ", +" | sfdisk -N 2 -f /dev/vda
+partprobe
+pvresize /dev/vda2
+lvresize fedora/root -l+100%FREE -r
```
Generalization that also works on RHEL:
```
rootlv=$(findmnt --noheadings -oSOURCE /)
lvresize $rootlv -l+100%FREE -r
```

jikortus commented on test/composertest.py:
```diff
@@ -0,0 +1,59 @@
+class ComposerTestCase(unittest.TestCase):
+    image = testvm.DEFAULT_IMAGE
+
+    def setUp(self):
```
Some of the current tests need some sort of specific setup (changing the existing files used by composer) - I think this could be the right place to do this setup? So far I can think of two cases - changing sshd configuration to accept login with empty passwords and modifying a blueprint to contain custom kernel commandline parameters - please see https://github.com/weldr/lorax/blob/master/tests/test_cli.sh#L7.

larskarlitski replied:
Yes and no :)

This `setUp` is called for all tests. Specific setup that should only happen for some of the tests should go into subclasses of this class (see test/check-cli for example, you could add a `setUp` there).

For now, the tests simply call out the `tests/cli/*.sh` scripts. We don't need to move anything over, unless it's easier to do in python.

atodorov replied:
FTR @jikortus means we do modify some of the blueprints/kickstart files to make it easier for testing. For example live-iso by default forbids SSH connections because root account doesn't have a password there. 

Since this PR has already been merged we can work on fixing the remaining issues one by one. However be mindful that QE uses the existing bash scripts as they are to validate RPM builds so they must continue to work in stand-alone mode. Which likely means duplicating setup in some places (yikes).

larskarlitski commented:
Update: increase the timeout for starting lorax to 6 minutes. 90 seconds was too little.

martinpitt commented:
Whee! Thanks @larskarlitski !
